src/python/grpcio_tests/tests_aio/unit/channel_argument_test.py

```diff
@@ -0,0 +1,168 @@
+# Copyright 2019 The gRPC Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests behavior around the Core channel arguments."""
+
+import asyncio
+import logging
+import unittest
+import socket
+
+import grpc
+import random
+
+from grpc.experimental import aio
+from src.proto.grpc.testing import messages_pb2
+from src.proto.grpc.testing import test_pb2_grpc
+from tests.unit.framework.common import test_constants
+from tests_aio.unit._test_server import start_test_server
+from tests_aio.unit._test_base import AioTestBase
+# 100 servers in sequence
+
+_RANDOM_SEED = 42
+
+_ENABLE_REUSE_PORT = 'SO_REUSEPORT enabled'
+_DISABLE_REUSE_PORT = 'SO_REUSEPORT disabled'
+_SOCKET_OPT_SO_REUSEPORT = 'grpc.so_reuseport'
+_OPTIONS = (
+    (_ENABLE_REUSE_PORT, ((_SOCKET_OPT_SO_REUSEPORT, 1),)),
+    (_DISABLE_REUSE_PORT, ((_SOCKET_OPT_SO_REUSEPORT, 0),)),
+)
+
+_NUM_SERVER_CREATED = 100
+
+_GRPC_ARG_MAX_RECEIVE_MESSAGE_LENGTH = 'grpc.max_receive_message_length'
+_MAX_MESSAGE_LENGTH = 1024
+
+
+class _TestPointerWrapper(object):
+
+    def __int__(self):
+        return 123456
+
+
+_TEST_CHANNEL_ARGS = (
+    ('arg1', b'bytes_val'),
+    ('arg2', 'str_val'),
+    ('arg3', 1),
+    (b'arg4', 'str_val'),
+    ('arg6', _TestPointerWrapper()),
+)
+
+_INVALID_TEST_CHANNEL_ARGS = [
+    {
+        'foo': 'bar'
+    },
+    (('key',),),
+    'str',
+]
+
+
+async def test_if_reuse_port_enabled(server: aio.Server):
+    port = server.add_insecure_port('127.0.0.1:0')
+    await server.start()
+
+    try:
+        another_socket = socket.socket(family=socket.AF_INET)
+        another_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        another_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
+        another_socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, True)
+        another_socket.bind(('127.0.0.1', port))
+    except OSError as e:
+        assert 'Address already in use' in str(e)
+        return False
+    else:
+        return True
+    finally:
+        another_socket.close()
+
+
+class TestChannelArgument(AioTestBase):
+
+    async def setUp(self):
+        random.seed(_RANDOM_SEED)
+
+    @unittest.skip('https://github.com/grpc/grpc/issues/20667')
+    async def test_server_so_reuse_port_is_set_properly(self):
+
+        async def test_body():
+            fact, options = random.choice(_OPTIONS)
```

Comment:
Very nice! I love this style of test! :heart:
The level of coverage is so much higher. The only question is how long this test takes to complete.

Reply:
The length of the test is less than 1 second. Tested with a hacky solution that only work for Python 3.7, but not for Python 3.6 due to coroutine execution order changed.